### PR TITLE
EIP-712 empty array handling fix

### PR DESCRIPTION
**Pull request by apaillier-ledger**
## Description

* Fix wrong signatures with empty array within nested structs
* Now uses Web3.py to verify EIP-712 signatures in Ragger tests

Fixes https://github.com/LedgerHQ/ledger-live/issues/6065

## Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [x] Tests
- [ ] Documentation
- [ ] Other (for changes that might not fit in any category)

### CODE REVIEW

File: src_features/signMessageEIP712/path.c [hunk 1]
```diff
@@ -285,23 +286,33 @@ static bool array_depth_list_pop(void) {
...
+    const void *starting_field_ptr, *field_ptr;
     const char *typename;
     uint8_t typename_len;
     uint8_t hash[KECCAK256_HASH_BYTESIZE];
 
     if (path_struct == NULL) {
         return false;
     }
-    if ((field_ptr = get_field(NULL)) == NULL) {
+    if ((starting_field_ptr = get_field(NULL)) == NULL) {
```
Comment (fbeutin-ledger) [inline src_features/signMessageEIP712/path.c:304]: split assignment and check

File: src_features/signMessageEIP712/path.c [hunk 2]
```diff
@@ -285,23 +286,33 @@ static bool array_depth_list_pop(void) {
...
         return false;
     }
+    field_ptr = starting_field_ptr;
     while (struct_field_type(field_ptr) == TYPE_CUSTOM) {
```
Comment (fbeutin-ledger) [inline src_features/signMessageEIP712/path.c:308]: add a comment to explain what this loop is looking for

File: src_features/signMessageEIP712/path.c [hunk 3]
```diff
@@ -313,11 +324,15 @@ static bool path_update(void) {
...
         }
-        // get the struct typehash
-        if (type_hash(typename, typename_len, hash) == false) {
-            return false;
+
+        // The only times they are both at false is if we are traversing an empty array
+        if ((skip_if_array != false) || (stop_at_array != false)) {
+            // get the struct typehash
+            if (type_hash(typename, typename_len, hash) == false) {
+                return false;
```
Comment (fbeutin-ledger) [inline src_features/signMessageEIP712/path.c:332]: explain better what this get is trying to assert and why it's an error

File: src_features/signMessageEIP712/path.c (hunk 2 above)
Comment (apaillier-ledger) [reply src_features/signMessageEIP712/path.c:309]: Already explained in the function's comment :+1:

**Review by fbeutin-ledger — APPROVED** [review]

**Pull request merged**